include/bluetooth/gatt.h

```diff
@@ -100,44 +100,55 @@ enum {
+struct bt_gatt_attr;
+
+/** @typedef bt_gatt_attr_read_func_t
+ *  @brief Attribute read callback
+ *
+ *  The callback can also be used locally to read the contents of the
+ *  attribute in which case no connection will be set.
+ *
+ *  @param conn   The connection that is requesting to read
+ *  @param attr   The attribute that's being read
+ *  @param buf    Buffer to place the read result in
+ *  @param len    Length of data to read
+ *  @param offset Offset to start reading from
+ *
+ *  @return Number of bytes read, or in case of an error
+ *          BT_GATT_ERR() with a specific ATT error code.
+ */
+typedef ssize_t (*bt_gatt_attr_read_func_t)(struct bt_conn *conn,
+					    const struct bt_gatt_attr *attr,
+					    void *buf, uint16_t len,
+					    uint16_t offset);
+
+/** @typedef bt_gatt_attr_write_func_t
+ *  @brief Attribute write callback
+ *
+ *  @param conn   The connection that is requesting to write
+ *  @param attr   The attribute that's being written
+ *  @param buf    Buffer with the data to write
+ *  @param len    Number of bytes in the buffer
+ *  @param offset Offset to start writing from
+ *  @param flags  Flags (BT_GATT_WRITE_*)
+ *
+ *  @return Number of bytes written, or in case of an error
+ *          BT_GATT_ERR() with a specific ATT error code.
+ */
+typedef ssize_t (*bt_gatt_attr_write_func_t)(struct bt_conn *conn,
+					     const struct bt_gatt_attr *attr,
+					     const void *buf, uint16_t len,
+					     uint16_t offset, uint8_t flags);
+
 /** @brief GATT Attribute structure. */
 struct bt_gatt_attr {
 	/** Attribute UUID */
 	const struct bt_uuid *uuid;
-
-	/** @brief Attribute read callback
-	 *
-	 *  The callback can also be used locally to read the contents of the
-	 *  attribute in which case no connection will be set.
-	 *
-	 *  @param conn   The connection that is requesting to read
-	 *  @param attr   The attribute that's being read
-	 *  @param buf    Buffer to place the read result in
-	 *  @param len    Length of data to read
-	 *  @param offset Offset to start reading from
-	 *
-	 *  @return Number fo bytes read, or in case of an error
-	 *          BT_GATT_ERR() with a specific ATT error code.
-	 */
-	ssize_t (*read)(struct bt_conn *conn, const struct bt_gatt_attr *attr,
-			void *buf, uint16_t len, uint16_t offset);
-
-	/** @brief Attribute write callback
-	 *
-	 *  @param conn   The connection that is requesting to write
-	 *  @param attr   The attribute that's being written
-	 *  @param buf    Buffer with the data to write
-	 *  @param len    Number of bytes in the buffer
-	 *  @param offset Offset to start writing from
-	 *  @param flags  Flags (BT_GATT_WRITE_*)
-	 *
-	 *  @return Number of bytes written, or in case of an error
-	 *          BT_GATT_ERR() with a specific ATT error code.
-	 */
-	ssize_t	(*write)(struct bt_conn *conn, const struct bt_gatt_attr *attr,
-			 const void *buf, uint16_t len, uint16_t offset,
-			 uint8_t flags);
-
+	/** Attribute read callback */
```

Comment:
Personally, I am happy to have the callback function definitions within the struct, so I do not see it as necessary to move them out.

Reply:
My idea was to allow easier reuse of the function (pointer) types without having to copy the definitions to the user code, plus easier reference from other parts of Doxygen docs. You would be the second person preferring them inside the struct, so if these arguments don't sway you, I'll leave them in place :slightly_smiling_face: